src/Libraries/Microsoft.Extensions.AI.Abstractions/Contents/JsonFunctionCallUtilities.cs

```diff
@@ -0,0 +1,448 @@
+
+        FunctionParameterKey key = (
+            type,
+            parameterName,
+            description,
+            hasDefaultValue,
+            defaultValue,
+            IncludeSchemaUri: false,
+            inferenceOptions.DisallowAdditionalProperties,
+            inferenceOptions.IncludeTypeInEnumSchemas);
+
+        return GetJsonSchemaCached(serializerOptions, key);
+    }
+
+    /// <summary>Infers a JSON schema for the specified type.</summary>
+    /// <param name="type">The type for which to generate the schema.</param>
+    /// <param name="serializerOptions">The options used to extract the schema from the specified type.</param>
+    /// <param name="description">The description of the parameter.</param>
+    /// <param name="hasDefaultValue">Whether the parameter is optional.</param>
+    /// <param name="defaultValue">The default value of the optional parameter, if applicable.</param>
+    /// <param name="inferenceOptions">The options controlling schema inference.</param>
+    /// <returns>A <see cref="JsonElement"/> representing the schema.</returns>
+    public static JsonElement InferJsonSchema(
+        Type? type,
+        JsonSerializerOptions serializerOptions,
+        string? description = null,
+        bool hasDefaultValue = false,
+        object? defaultValue = null,
+        JsonSchemaInferenceOptions? inferenceOptions = null)
+    {
+        _ = Throw.IfNull(serializerOptions);
+        inferenceOptions ??= JsonSchemaInferenceOptions.Default;
+
+        FunctionParameterKey key = (
+            type,
+            ParameterName: null,
+            description,
+            hasDefaultValue,
+            defaultValue,
+            inferenceOptions.IncludeSchemaKeyword,
+            inferenceOptions.DisallowAdditionalProperties,
+            inferenceOptions.IncludeTypeInEnumSchemas);
+
+        return GetJsonSchemaCached(serializerOptions, key);
+    }
+
+    private static JsonElement GetJsonSchemaCached(JsonSerializerOptions options, FunctionParameterKey key)
+    {
+        options.MakeReadOnly();
+        ConcurrentDictionary<FunctionParameterKey, JsonElement> cache = _schemaCaches.GetOrCreateValue(options);
+
+        if (cache.Count >= CacheSoftLimit)
+        {
+            return GetJsonSchemaCore(options, key);
+        }
+
+        return cache.GetOrAdd(
+            key: key,
+#if NET
+            valueFactory: static (key, options) => GetJsonSchemaCore(options, key),
+            factoryArgument: options);
+#else
+            valueFactory: key => GetJsonSchemaCore(options, key));
+#endif
+    }
+
+    private static JsonElement GetJsonSchemaCore(JsonSerializerOptions options, FunctionParameterKey key)
+    {
+        _ = Throw.IfNull(options);
+        options.MakeReadOnly();
+
+        if (options.ReferenceHandler == ReferenceHandler.Preserve)
+        {
+            throw new NotSupportedException("Schema generation not supported with ReferenceHandler.Preserve enabled.");
```

Comment:
Can you add a comment about why this is the case?

Reply:
You added this :-) This is replicating a check in JsonSchemaExporter/JsonSchemaMapper so in many ways it's redundant. I'm interpreting this as an attempt to throw the same error regardless of whether the AI metadata is defining a type, but if not we can just remote it.

Reply:
I didn't "add" it, I just copied it from https://github.com/eiriktsarpalis/stj-schema-mapper/blob/3cc062d1c4711fc83626fb59a5cf2d96cb76bc7c/src/JsonSchemaMapper/JsonSchemaMapper.cs#L410.

Reply:
Gotcha. We can remove it given that the check is made by `JsonSchemaExporter` as well.